File: packages/bash-wasm/sandboxes/js/sandbox.ts
import { task } from '@capsule-run/sdk';
import path from 'path';
import fs from 'fs';
import fsPromises from 'fs/promises';

import type { State } from '@capsule-run/bash-types';

function wasmRelative(cwd: string, filePath: string): string {
  return path.resolve(cwd, filePath).replace(/^\//, '');
}

function resolveNodeModule(fromPath: string, id: string): string | null {
  let dir = path.dirname('/' + fromPath);

  while (true) {
    const base = wasmRelative('/', path.join(dir, 'node_modules', id));
    const candidates = [base, base + '.js', base + '/index.js'];

    const pkgJson = base + '/package.json';

    if (fs.existsSync(pkgJson)) {
      try {
        const pkg = JSON.parse(fs.readFileSync(pkgJson, 'utf-8') as string);
        const main = pkg.main || 'index.js';
        candidates.unshift(wasmRelative('/', path.join(dir, 'node_modules', id, main)));
      } catch {}
    }

    for (const candidate of candidates) {
      if (fs.existsSync(candidate)) return candidate;
    }

    const parent = path.dirname(dir);
    if (parent === dir) return null;

    dir = parent;
  }
}

const builtins: Record<string, unknown> = { fs, path, 'fs/promises': fsPromises };

const makeRequire = (fromPath: string) => (id: string) => {
  let depPath: string;

  if (!id.startsWith('.') && !id.startsWith('/')) {
    if (id in builtins) return builtins[id];

    const resolved = resolveNodeModule(fromPath, id);

    if (!resolved) throw new Error(`Cannot find module '${id}'`);

    depPath = resolved;
  } else {
    const base = wasmRelative('/', path.resolve('/' + path.dirname(fromPath), id));

    depPath = fs.existsSync(base)
      ? base
      : fs.existsSync(base + '.js')
        ? base + '.js'
        : base + '/index.js';
  }

  const src = fs.readFileSync(depPath, 'utf-8') as string;
  const m = { exports: {} as any };
  const fn = new Function('module', 'exports', 'require', '__filename', '__dirname', src);

  fn(m, m.exports, makeRequire(depPath), depPath, path.dirname(depPath));

  return m.exports;
};

const executeFile = task(
  { name: 'executeFile', compute: 'MEDIUM', ram: '512MB', allowedHosts: ['*'] },
  async (state: State, filePath: string, args: string[]) => {
    const capturedOutput: string[] = [];
    const relPath = wasmRelative(state.cwd, filePath);

    process.chdir(state.cwd);

    Object.defineProperty(process, 'argv', {
      value: ['node', relPath, ...args],
      writable: true,
      configurable: true,
    });

    const capture = (...logArgs: any[]): void => {
      capturedOutput.push(logArgs.map((a) => String(a)).join(' '));
    };

    const originalLog = console.log;
    const originalError = console.error;
    const originalWarn = console.warn;
    console.log = capture;
    console.error = capture;
    console.warn = capture;

    try {
      const code = fs.readFileSync(relPath, 'utf-8') as string;
      const mod = { exports: {} as any };
      const customRequire = makeRequire(relPath);

      const fn = new Function('module', 'exports', 'require', '__filename', '__dirname', code);
      fn(mod, mod.exports, customRequire, relPath, path.dirname(relPath));

      const output = capturedOutput.join('\n');

      if (output) return output;

      return Object.keys(mod.exports).length > 0 ? mod.exports : null;
    } finally {
      console.log = originalLog;
      console.error = originalError;
      console.warn = originalWarn;
    }
  },
);

const executeCode = task(
  { name: 'executeCode', compute: 'LOW', ram: '256MB', allowedHosts: ['*'] },
  async (state: State, code: string): Promise<unknown> => {
    process.chdir(state.cwd);
    const capturedOutput: string[] = [];

    const capture = (...args: any[]): void => {
      capturedOutput.push(args.map((arg) => String(arg)).join(' '));
    };

    const originalLog = console.log;
    const originalError = console.error;
    const originalWarn = console.warn;
    console.log = capture;
    console.error = capture;
    console.warn = capture;

    const require = makeRequire(wasmRelative(state.cwd, '.'));

    try {
      let result;
      try {
        const wrapper = new Function('require', 'console', `return eval(${JSON.stringify(code)})`);
        result = await wrapper(require, console);
      } catch (e) {
        if (e instanceof SyntaxError) {
          const fn = new Function('require', 'console', code);
          result = await fn(require, console);
        } else {
          throw e;
        }
      }

      const output = capturedOutput.join('\n');

      if (output) {
        return result != null ? output + '\n' + String(result) : output;
      }

      return result ?? null;
    } finally {
      console.log = originalLog;
      console.error = originalError;
      console.warn = originalWarn;
    }
  },
);

export const resolvePath = task(
  { name: 'resolvePath', compute: 'LOW', ram: '32MB' },
  async (state: State, targetPath: string) => {
    process.chdir(state.cwd);

    if (!fs.existsSync(targetPath) && targetPath !== state.cwd) {
      throw new Error(`Path ${targetPath} does not exist`);
    }

    return path.resolve(targetPath);
  },
);

export const main = task(
  { name: 'main', compute: 'HIGH' },
  async (action: string, state: string, ...args: string[]): Promise<unknown> => {
    let parsedArgs = args.filter((arg) => typeof arg !== 'object'); // remove the unused kwargs
    let response: {
      success: boolean;
      result: unknown;
      error: { error_type: string; message: string } | null;
    };

    let parsedState: State = JSON.parse(state);

    if (action === 'PRELOAD') {
      response = { success: true, result: 'Sandbox preloaded successfully', error: null };
    } else if (action === 'EXECUTE_CODE') {
      response = await executeCode(parsedState, parsedArgs[0]);
    } else if (action === 'EXECUTE_FILE') {
      response = await executeFile(parsedState, parsedArgs[0], parsedArgs.slice(1));
    } else if (action === 'RESOLVE_PATH') {
      response = await resolvePath(parsedState, parsedArgs[0]);
    } else {
      throw new Error(`Invalid action: ${action}`);
    }

    if (!response.success && response.error) {
      throw new Error(response.error.message);
    }

    return response.result;
  },
);

File: packages/bash-wasm/sandboxes/python/sandbox.py
import ast
import sys
import os
import json
import builtins
from io import StringIO
from dataclasses import dataclass
from capsule import task

@dataclass
class State:
    cwd: str
    env: dict
    lastExitCode: int

    @staticmethod
    def from_json(raw: str) -> "State":
        data = json.loads(raw)
        return State(
            cwd=data["cwd"],
            env=data["env"],
            lastExitCode=data["lastExitCode"],
        )


def wasm_relative(cwd: str, file_path: str) -> str:
    joined = os.path.normpath(os.path.join(cwd, file_path))
    return joined.lstrip("/")


@task(name="executeFile", compute="MEDIUM", ram="512MB", allowed_hosts=["*"])
def execute_file(state: str, file_path: str, args: list[str]):
    state = State.from_json(state)
    rel_path = wasm_relative(state.cwd, file_path)
    file_dir = os.path.dirname(rel_path) or "."

    site_packages = os.path.join(state.cwd, "site-packages")
    if site_packages not in sys.path:
        sys.path.insert(0, site_packages)

    captured_output = StringIO()
    old_stdout = sys.stdout
    old_argv = sys.argv
    old_path = sys.path[:]
    sys.stdout = captured_output
    sys.argv = [rel_path] + list(args)
    sys.path.insert(0, file_dir)

    try:
        with open(rel_path, 'r') as f:
            source = f.read()
        global_env = {
            "__builtins__": builtins,
            "__name__": "__main__",
            "__file__": rel_path,
        }
        exec(compile(source, rel_path, 'exec'), global_env)
    finally:
        sys.stdout = old_stdout
        sys.argv = old_argv
        sys.path[:] = old_path

    output = captured_output.getvalue()
    public_result = {}
    for k, v in global_env.items():
        if k.startswith("__"):
            continue
        try:
            json.dumps(v)
            public_result[k] = v
        except (TypeError, ValueError):
            pass

    if output:
        return output.rstrip("\n") + "\n" + json.dumps(public_result) if public_result else output.rstrip("\n")

    return public_result if public_result else None


@task(name="executeCode", compute="LOW", ram="256MB", allowed_hosts=["*"])
def execute_code(state: str, code: str):
    state = State.from_json(state)

    site_packages = os.path.join(state.cwd, "site-packages")
    if site_packages not in sys.path:
        sys.path.insert(0, site_packages)

    tree = ast.parse(code)

    if not tree.body:
        return None

    last_node = tree.body[-1]
    local_env = {"__builtins__": builtins}

    captured_output = StringIO()
    old_stdout = sys.stdout
    sys.stdout = captured_output

    try:
        if isinstance(last_node, ast.Expr):
            tree.body.pop()
            if tree.body:
                exec(compile(tree, filename="<ast>", mode="exec"), local_env)
            result = eval(
                compile(ast.Expression(last_node.value), filename="<ast>", mode="eval"),
                local_env,
            )
        else:
            exec(compile(tree, filename="<ast>", mode="exec"), local_env)
            result = local_env.get("result")
    finally:
        sys.stdout = old_stdout

    output = captured_output.getvalue()

    if output:
        return output.rstrip("\n") + ("\n" + str(result) if result is not None else "")

    return result


@task(name="main", compute="HIGH")
def main(action: str, state: str, *args: str):
    if action == "PRELOAD":
        response = {"success": True, "result": "Sandbox preloaded successfully", "error": None}
    elif action == "EXECUTE_CODE":
        response = execute_code(state, args[0])
    elif action == "EXECUTE_FILE":
        response = execute_file(state, args[0], list(args[1:]))
    else:
        raise ValueError(f"Invalid action: {action}")

    if isinstance(response, dict):
        if not response.get("success"):
            error = response.get("error")
            if isinstance(error, dict):
                raise Exception(error.get("message", str(error)))
            raise Exception(str(error))
        if response.get("result") is not None:
            return response["result"]

    return response
